xds/reloading_stream.go

```diff
@@ -0,0 +1,108 @@
+package xds
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	v3discoverypb "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"google.golang.org/grpc"
+
+	"github.com/gojek/courier-go/xds/backoff"
+	"github.com/gojek/courier-go/xds/log"
+)
+
+type adsStream v3discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient
+
+type reloadingStream struct {
+	mu       sync.RWMutex
+	reloadWG sync.WaitGroup
+	log      log.Logger
+	s        adsStream
+	ns       func() v3discoverypb.AggregatedDiscoveryServiceClient
+
+	strategy    backoff.Strategy
+	onReConnect func(error)
+	reloadCh    chan error
+	connTimeout time.Duration
+}
+
+func (r *reloadingStream) Send(req *v3discoverypb.DiscoveryRequest) error {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	return r.s.Send(req)
+}
+
+func (r *reloadingStream) Recv() (*v3discoverypb.DiscoveryResponse, error) {
+	resp, err := r.safeRead()
+	if err == nil {
+		return resp, nil
+	}
+
+	r.reloadWG.Add(1)
+	r.reloadCh <- err
+	r.reloadWG.Wait()
+
+	return r.safeRead()
```

Comment:
As we discussed, should we just safeRead once and send error on reloadCh on receiving an error?